Adds asSet() to Bag for #726

victornoel opened the pull request:
:warning: **Edit: all of this is outdated, see #869 for the latest design** :warning: 

Finally had the time to finish my PR for #726.

I made the following design decisions:
- The different `Bag` interfaces exposes `asSet()` which is a view on the unique elements of the bag
- They are either a `SetIterable` or a `MutableSet`
- I didn't tackle returning sorted o immutable variants because
 - implementing it was way too complex
 - this is not the primary use case for accessing a `Bag` as a `Set` IMHO
- The synchronized, unmodifiable and multi reader variants are implemented though
- Only for `ImmutableArrayBag` and `ImmutableSortedBagImpl`, I had to implement it by subclassing `AbstractImmutableSet` inline: since they are both based on arrays, I only implemented the minimum because I don't think it would have much impact on performance to do more.

One open question is: should the inline subclassing be extracted in a separate class that wraps an array maybe?

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/AbstractHashBag.java:
```diff
@@ -373,4 +375,10 @@ public void remove()
+    @Override
+    public MutableSet<T> asSet()
+    {
+        return SetAdapter.adapt(this.items.keySet());
```
```java
@Test
public void asSet()
{
    MutableBag<Integer> bag = Bags.mutable.with(1, 2, 2, 3, 3, 3);
    MutableSet<Integer> set = bag.asSet();
    set.remove(3);
    Verify.assertSize(3, bag);
}
```

```
java.lang.AssertionError: Incorrect size for iterable; 
Expected :3
Actual   :6
```

victornoel replied:
Ah, indeed, I didn't think about that at all! So yes, let's do that and stop trying to handle the mutable case. This can be done in another iteration I suppose

victornoel replied:
One caveat with this is that if I do that, the return type of `MutableBag.asSet` is not a `MutableSet` anymore and I can't easily wrap all of the variants such as synchronized, unmodifiable and multireader without using some potentially unreliable casting...

Could we start by declaring `asSet` as returning a `MutableSet` but it would be an `UnmodifiableMutableSet` for now?

motlin replied:
SetIterable is more generic. You don't have to change anything about the implementations except that they also have to be unmodifiable.

victornoel replied:
Sorry, I wasn't clear, my problem is in those kind of place: https://github.com/eclipse/eclipse-collections/pull/797/files#diff-bf399276403945a5c7592cb17e78051fR1081

As you can see, to be able to use `MultiReaderUnifiedSet.of(MutableSet)`, I need that `this.getDelegate().asSet()` returns a `MutableSet`, but in this case, the delegate is a `MutableBag`. So if `MutableBag.asSet()` returns only a `SetIterable`, I have a typing problem that I can only solve via casting, which is dangerous because I'm not sure that the implementation will actually return a `MutableSet`.

Do you see what I mean?

victornoel replied:
In my last push, I've used the most precise interface possible everywhere except for the fact that `MutableSet` all returns unmodifiable ones in practice until the invariants can be fixed

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableArrayBag.java:
```diff
@@ -530,4 +535,68 @@ public void remove()
+            public int hashCode()
+            {
+                return Arrays.hashCode(ImmutableArrayBag.this.keys);
+            }
```
> Consider pulling up equals() and hashCode() onto AbstractImmutableSet.
How would that work? AbstractImmutableSet does not know how the set is actually implemented: each of the subclasses is implementing it depending on their content.

motlin replied:
All Sets follow the same contract though. Note that AbstractSet has an implementation of equals() and hashCode().

motlin replied:
And this one has the implementation of equals() that can be pulled up.

victornoel replied:
Ok, I can do it, I'm not sure it will be great in terms of performance but why not :)

victornoel replied:
(for equals, it's clear that it is a good idea though)

victornoel replied:
So any reason not to bring `hashCode` implementation in `AbstractRichIterable`?

victornoel replied:
Ah, the only problem is that checkstyles wants both `hashCode` and `equals` implemented together, so it's all or nothing, so if I bring both `equals` and `hashCode` to `AbstractImmutableSet` with `hashCode` being suboptimized, then if I want to override it in the subclasses, I also need to copy `equals` and in that case, it makes no sense to bring also into `AbstractImmutableSet`. Or I just don't override `hashCode` but it's sad...

victornoel replied:
Actually, apparently, checkstyle only wants that equals is declared if `hashCode` is declared, not the opposite.

And I can always use `super` in any case :)

victornoel replied:
I've pushed some changes that does this + some improvements around equals in set implementations (in separate commits for easier review)

victornoel commented on eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/MutableSortedBag.java:
```diff
@@ -271,4 +272,7 @@
     MutableSortedBag<T> drop(int count);
+
+    @Override
+    MutableSet<T> asSet();
```
I tried but the main problem is coming from `MutableSortedMap.keySet` (the underlying map in `TreeBag`) which returns only a `MutableSet` with the following comment. So if something can be done about that, then yes this one could be sorted IMHO.
```
    /**
     * The underlying set for the keys is sorted in ascending order according to their natural ordering or a custom comparator.
     * However, Java 5 TreeMap returns a keySet that does not inherit from SortedSet therefore we have decided to
     * return the keySet simply as a MutableSet to maintain Java 5 compatibility.
     */
    //todo: Change return type to MutableSortedSet when we move to Java 6
```

victornoel replied:
I've pushed some changes where I simply cast the keySet to bypass this, should I open an issue to solve the todo above?

victornoel replied:
Actually I can't because `MutableSortedSet` already wrapped it in `SetAdapter` unfortunately...

victornoel replied:
And we can't simply wrap the keySet instead in SortedSetAdapter, because `MutableSortedSet` is not a subclass of `MutableSet` but of `MutableSetIterable`

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableArrayBag.java:
```diff
@@ -530,4 +535,68 @@ public void remove()
+                return ImmutableArrayBag.this.keys.length > 0
+                    ? ImmutableArrayBag.this.keys[ImmutableArrayBag.this.keys.length - 1]
+                    : null;
+            }
```
Totally

victornoel replied:
Done in my latest push

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableHashBag.java:
```diff
@@ -677,4 +679,10 @@ protected Object writeReplace()
+    public SetIterable<T> asSet()
+    {
+        return UnmodifiableMutableSet.of(this.delegate.asSet());
+    }
```
It's because implementing an actual immutable version of this view while the underlying bad is itself mutable is a lot of work in this class.

Since I only needed to return a `SetIterable`, I simply protected the integrity of the bag by wrapping it in an unmodifiable decorator.

This would be different if we wanted to return an `ImmutableSet` here...

motlin replied:
In some places you do return MutableSet though, right? Should we return SetIterable everywhere?

victornoel replied:
Yes, as I explained in the PR description and in https://github.com/eclipse/eclipse-collections/pull/797#discussion_r379922824, I decided to return either `SetIterable` or `MutableSet`.

But we could decide to be more precise and return also `ImmutableSet`, `ImmutableSortedSet`, MutableSortedSet` and so on when possible. It's not always easy, see for example https://github.com/eclipse/eclipse-collections/pull/797#discussion_r379891931 but it could be done I suppose.

victornoel replied:
In my last push, I've used the most precise interface possible everywhere except for the fact that `MutableSet` all returns unmodifiable ones in practice until the invariants can be fixed

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/ImmutableArrayBag.java:
```diff
@@ -530,4 +535,68 @@ public void remove()
+    @Override
+    public SetIterable<T> asSet()
+    {
+        return new AbstractImmutableSet<T>()
```
Good point

victornoel replied:
Done in my latest push, but not static

victornoel commented:
@motlin any feedbacks on the best way to tackle the problems discussed above?

motlin commented:
@victornoel my recommendation would still be to just use `SetIerable` for now. That way the interface doesn't promise to return a MutableSet in places where you'd expect to be able to mutate but we can't yet support it. It's a step in the right direction and we can always support mutation in another major version.

victornoel commented:
@motlin yes, I will go with that then, thanks

victornoel commented:
@motlin one more advice request: if we use `SetIterable` everywhere, it becomes difficult to take advantage of existing `Synchronized` and `Unmodifiable` wrappers for mutable sets, what do you recommend? Writing some for `SetIterable` itself?

victornoel commented:
Okay, so I pushed a final version with:
- read only sets as result of `asSet`:
  - `SetIterable` for all various `MutableBag`
  - corresponding `Immutable` and `Sorted` and so on for all the non mutable bags.
- for multireader and synchronized bags, I had to add `SynchronizedSetIterable` and `SynchronizedParallelSetIterable` classes

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/immutable/AbstractImmutableSet.java:
```diff
@@ -474,4 +474,37 @@ public boolean isProperSubsetOf(SetIterable<? extends T> candidateSuperset)
+        {
+            return false;
+        }
+    }
```
Similarly out of place

victornoel replied:
@motlin ok, and anyway findbugs was complaining `hashCode` to be missing, I moved it in the one subclass that delegated to super and that also implemented `hashCode`.

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/immutable/ImmutableUnifiedSet.java:
```diff
@@ -43,6 +44,16 @@ public int size()
+        {
+            return false;
+        }
+
```
Similarly out of place

victornoel replied:
@motlin I'm not clear why for this one: it's in a concrete class and with an implementation of `hashCode` next to it.

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/MultiReaderUnifiedSet.java:
```diff
@@ -131,6 +131,16 @@ private MultiReaderUnifiedSet(MutableSet<T> newDelegate, ReadWriteLock newLock)
+    public static <T> MultiReaderUnifiedSet<T> of(MutableSet<T> newDelegate, ReadWriteLock newLock)
+    {
+        return new MultiReaderUnifiedSet<>(newDelegate, newLock);
+    }
```
These are non-copy constructors. We have to remove them, or at least hide their visibility.

victornoel replied:
@motlin good point, actually I added them in the past but don't use them, so I will just remove them.

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/MultiReaderUnifiedSet.java:
```diff
@@ -433,6 +443,16 @@ public MutableShortSet collectShort(ShortFunction<? super T> shortFunction)
+        if (!(o instanceof Set))
+        {
+            return false;
+        }
```
This seems like a clever optimization. How did it come up?

victornoel replied:
@motlin I think I tried to be consistent with the other set equals and it seemed that since it was anyway a requirement that all and every set implementations (even immutable in particular) were required to implement Set for equality test to work in both direction, I thought I would add that.
Maybe this is incorrect though and we would want those to be comparable also with non-set?

victornoel replied:
@motlin it was already there for example for `UnifiedSet`.

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/SetAdapter.java:
```diff
@@ -152,6 +152,16 @@ public boolean containsAll(Collection<?> collection)
+        {
+            return false;
+        }
+
```
This seems out of place.

victornoel replied:
@motlin why is that? Maybe we don't want to test for being in instance of `Set` and let `AbstractCollectionAdapter` define `equals` directly?

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/UnmodifiableMutableSet.java:
```diff
@@ -104,6 +104,16 @@
+        {
+            return false;
+        }
+
```
Seems out of place

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/sorted/immutable/AbstractImmutableSortedSet.java:
```diff
@@ -524,4 +524,35 @@ public int detectLastIndex(Predicate<? super T> predicate)
+        {
+            return false;
+        }
+    }
```
Why does this catch exceptions? Can we have hashcode here too?

victornoel replied:
@motlin some of the implementations of equals that rely on `Collection.containsAll` do catch those exception because `Collection.containsAll` document them and if they are thrown, it's not an error, it means the element are not contained (because of different types or things like that).

I can add `hashCode`, I didn't because I expected that all subclass should be more relevant to implement it, but if I add it, I'm not sure what is the best default implementation? An `hashCode` of all elements?

victornoel replied:
@motlin concerning hashCode, actually findbugs is complaining about it, so I moved it to the two subclasses that needed it.

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/sorted/mutable/SortedSetAdapter.java:
```diff
@@ -173,6 +173,16 @@ public boolean containsAll(Collection<?> collection)
+            return false;
+        }
+
         return this.delegate.equals(obj);
```
Seems out of place

victornoel replied:
@motlin what do you mean? Should I move `equals` and `hashCode` to AbstractCollectionAdapter maybe? But I wouldn't be able to test for instance of `Set` though

motlin commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/sorted/mutable/SynchronizedSortedSet.java:
```diff
@@ -98,7 +98,7 @@
      * with the collection.
      */
-    public static <E, S extends SortedSet<E>> MutableSortedSet<E> of(S set, Object lock)
+    public static <E, S extends SortedSet<E>> SynchronizedSortedSet<E> of(S set, Object lock)
```
Why is the return type changed?

victornoel replied:
@motlin I think I needed it in my first round of changes but don't anymore, I can revert. I left it because it was more precise and not breaking.

victornoel commented on eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/ImmutableBag.java:
```diff
@@ -219,4 +219,7 @@
     ImmutableList<ObjectIntPair<T>> bottomOccurrences(int count);
+
+    @Override
+    ImmutableSet<T> asSet();
```
@motlin as explained above, I decided to go for the most specific when possible: why restrict ourselves if we can be specific? For all non-mutable `Bag`, there is not problem with doing so. But I can revert of course if preferred.

victornoel commented on eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/ImmutableSortedBag.java:
```diff
@@ -257,4 +257,7 @@
     ImmutableSortedBag<T> drop(int count);
+
+    @Override
+    ImmutableSortedSet<T> asSet();
```
@motlin your choice :)

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBag.java:
```diff
@@ -438,4 +439,10 @@ protected Object readResolve()
+    @Override
+    public SetIterable<T> asSet()
+    {
+        return this.getMutableBag().asSet();
```
@motlin I can't call `asUnmodifiable()` because the return type of `asSet()` here is `SetIterable`: I am expecting it to be unmodifiable if needed be. I will add a test to ensure it is the case though!

victornoel replied:
@motlin actually the test was already there :)

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/AbstractSynchronizedRichIterable.java:
```diff
@@ -105,6 +105,11 @@ protected Object getLock()
+        if (obj == this || obj == this.delegate)
+        {
+            return true;
+        }
```
@motlin haha, good question, I'm not sure, I was moving the equals implementation around to see where it best fit and this happened to come to my mind ^^

motlin commented:
I see the problem with the types. If the view has no way of supporting
mutation, then it's ok.

On Sat, Mar 14, 2020 at 2:10 PM Victor Noël <notifications@github.com>
wrote:

> *@victornoel* commented on this pull request.
> ------------------------------
>
> In
> eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBag.java
> <https://github.com/eclipse/eclipse-collections/pull/797#discussion_r392607573>
> :
>
> > @@ -438,4 +439,10 @@ protected Object readResolve()
>      {
>          return this.getMutableBag().selectUnique();
>      }
> +
> +    @Override
> +    public SetIterable<T> asSet()
> +    {
> +        return this.getMutableBag().asSet();
>
> @motlin <https://github.com/motlin> I can't call asUnmodifiable() because
> the return type of asSet() here is SetIterable: I am expecting it to be
> unmodifiable if needed be. I will add a test to ensure it is the case
> though!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/pull/797#discussion_r392607573>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB3UISK25LGG5O7AM2UEH3RHPB7RANCNFSM4KV2JJUA>
> .
>

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/parallel/set/SynchronizedParallelSetIterable.java:
```diff
@@ -0,0 +1,108 @@
+        {
+            return this.delegate.groupByEach(function);
+        }
+    }
```
@motlin yes I copied all of this from `SynchronizedParallelUnsortedSetIterable` and adapted it without thinking too much. Not that they don't return `ParallelIterable` but `SetMultimap` so I think that's why they don't call wrap... but why are they different, good question, it's how it is defined in `ParallelIterable`.

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/set/AbstractUnifiedSet.java:
```diff
@@ -425,4 +425,37 @@ public boolean isProperSubsetOf(SetIterable<? extends T> candidateSuperset)
+        {
+            return false;
+        }
+    }
```
@motlin where would this go then? In the subclasses?

victornoel replied:
@motlin anyway findbugs was complaining `hashCode` to be missing, I moved it in the two subclasses that also implement `hashCode`.

motlin commented:
Using inner class views makes me nervous, especially in the immutable collections. I was imagining a class like SetFromBagAdapter that's almost like SetAdapter, but implemented by delegating to a Bag. As long as it doesn't delegate to a MutableBag, we can have high confidence that it never allows mutation as well as unsynchronized access.

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/SynchronizedSortedBag.java:
```diff
@@ -734,4 +736,13 @@ public int detectLastIndex(Predicate<? super T> predicate)
+    {
+        synchronized (this.getLock())
+        {
+            return SynchronizedSetIterable.of(this.getDelegate().asSet(), this.getLock());
```
@motlin I think this one should be alright, `getLock()` returns an `Object`.

victornoel commented on eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/SynchronizedBag.java:
```diff
@@ -443,4 +445,13 @@ public MutableShortBag collectShort(ShortFunction<? super T> shortFunction)
+    {
+        synchronized (this.getLock())
+        {
+            return SynchronizedSetIterable.of(this.getDelegate().asSet(), this.getLock());
```
@motlin I think this one should be alright, `getLock()` returns an `Object`.

victornoel commented:
@motlin trying back to work on this, you said:
> Using inner class views makes me nervous, especially in the immutable collections. I was imagining a class like SetFromBagAdapter that's almost like SetAdapter, but implemented by delegating to a Bag. As long as it doesn't delegate to a MutableBag, we can have high confidence that it never allows mutation as well as unsynchronized access.

How would that work? In order to implement `SetFromBagAdapter` by delegating to a `Bag` I would need to be able to access the internal of the bag... or do you mean that it is created by passing to it the underlying datastructure used inside the bag? the array or the map, depending on the implementation?

victornoel commented:
@motlin I have pushed some changes that I think all together make sense, here is a rundown of the changes that could raise discussion and rationale for the choices I made:
- I extracted the internal classes into their own file as you proposed
  - `ImmutableSetAdapter` wraps any `UnsortedSetIterable` and is used by `ImmutableHashBag`
    - I tried to override what was needed to have best performances
  - `ImmutableSetFromArrayBagAdapter` wraps the array of a `ImmutableArrayBag` directly
    - this could also be moved to set package and called `ImmutableArraySetAdapter` or something like that but I feel like it's better if it stays hidden next to the class using it
  - same for `ImmutableSortedSetFromArrayBagAdapter`.
  - we can't just have a simple `ImmutableSetFromBagAdapter` because it would need `asSet()` to be implemented (and that would lead to risk of strange infinite loops :)
  - I'm not sure how the serialization stuffs work in there, so I may have made mistakes...
- `MutableSet.asSet()` returns `MutableBag`
  - they all are unmodifiable for now: we should create a new issue to tackle this later.
  - it is mainly needed to elegantly take advantage of all the `MutableSet` decorators for synchronized, unmodifiable, multi reader, etc without pain. I tried many times in the previous iterations to do without it, but there always was a problem somewhere so I think it's best to return `MutableSet` and for now not honour the mutability contract of it.
- `MutableSortedBag` does not return `MutableSortedSet`
  - because of `MutableSortedMap.keySet` contract
  - I added comments there for now but maybe there is something else to be done so that this can go on?
  - this is the only place where it would be needed to change the signatures of interface methods in the future
- `MultiReaderBag` returns `MultiReaderSet`
  - I added a `MultiReaderSetAdapter` that share most of its implementation with `MultiReaderUnifiedSet` (shared in in `AbstractMultiReaderSet`).
  - I'm not sure how the serialization stuffs work in there, so I may have made mistakes...

motlin commented:
> we can't just have a simple ImmutableSetFromBagAdapter because it would need asSet() to be implemented (and that would lead to risk of strange infinite loops :)

Can't it return `this`?

> MutableSet.asSet() returns MutableBag

Something sounds backwards in this sentence

> ImmutableSetAdapter wraps any UnsortedSetIterable

This isn't what I meant - I'll have to take a closer look. Why does the wrapper need access to the Bag internals?

victornoel commented:
@motlin 
> Can't it return this?

what I mean is that if you wrap a bag, to implement some of the set method, you need to have access to the underlying datatstructure, for example to implement iterator, and in our situation there are two cases when this happens:
- for `ImmutableArrayBag`, in this case you need access to the array
  - That's why I introduced `ImmutableSetFromArrayBagAdapter` and also `ImmutableSortedSetFromArrayBagAdapter`
- for ImmutableHashBag, in this case you can indeed have an hypothetical `ImmutableSetFromBagAdapter` that takes the hidden mutable `HashBag` and then call `asSet()` on it, but if you pass it `this`, it will go into infinite loop of course
  - That's why I introduced `ImmutableSetAdapter` that introduces no risks but allow to achieve the same objective in a more generic way.

>> MutableSet.asSet() returns MutableBag

> Something sounds backwards in this sentence

it was, I will fix my comment :)

motlin commented:
I tried to write a set-from-bag adapter to see how feasible it is. It's nearly possible, it would need one new method on Bags for a distinct keys view. Take a look at the iterator() method to see what I mean.

```java
public class SetFromBagAdapter<T>
        extends AbstractImmutableSet<T>
{
    private final Bag<T> bag;

    public SetFromBagAdapter(Bag<T> bag)
    {
        this.bag = bag;
    }

    @Override
    public int size()
    {
        return this.bag.sizeDistinct();
    }

    @Override
    public T getFirst()
    {
        return this.bag.getFirst();
    }

    @Override
    public T getLast()
    {
        return this.bag.getLast();
    }

    @Override
    public void each(Procedure<? super T> procedure)
    {
        this.bag.forEachWithOccurrences((each, parameter) -> procedure.accept(each));
    }

    @Override
    public Iterator<T> iterator()
    {
        // This works but is inefficient. It really creates the map in O(n) time and space.
        Iterator<T> iterator = this.bag.toMapOfItemToCount().keysView().iterator();

        // This doesn't work today, but it could without too much effort.
        return this.bag.keysView().iterator();
    }
}
```

motlin commented:
The in places like AbstractHashBag, it's pretty easy to implement the distinct view.

```java
    @Override
    public RichIterable<T> distinctView()
    {
        return this.items.keysView();
    }
```

victornoel commented:
@motlin if you look back at the issue #726 that spawned this whole PR, this `keyView` is the reason it exists :) so yes, a `keyView()` would be useful, but implementing means doing all that was done here.

motlin commented:
I did forget that context but I think this shows we can implement both with just a few lines of code?

victornoel commented:
@motlin no, because of all the other implementation of Bag that do not rely on a Map as in your example. For example `ImmutableArrayBag` is such an example. Let's also not forget that we want to have a O(1) method `asSet`, if I just want to be able to iterate on the distinct keys, as you said, I can always use `toMapOfItemToCount` but it's not efficient.

The purpose of this PR is to:
- implement such distinct key method, we decided in #726 that it was a `Set` view of the map
- that it behave in O(1), that's why the name of the method starting with `as`

Hence this proposal for `asSet`.

Honestly I've tried many alternative to arrive to this solution, with various iterations that can we witnessed here, I think we did go around all the possibilities and that the design proposed here is the best. Maybe to better understand that, it would make sense for you to play with the code and experience with the alternative you propose to either see why it doesn't work or to show me how it would work.

victornoel commented:
Closing this as this will most certainly never be merged and based on very outdated code
